### ISSUE TITLE
load tpm libraries into initramfs

### ISSUE THREAD

**Issue by glassman81** [tpm2-software/tpm2-tss#1509]
Since LUKS doesn't have official support for tpm2 yet, I'm using @AndreasFuchsSIT LUKS branch to try and encrypt my root ext4 partition. When I'm in initramfs, I can now attempt to use the cryptsetup luksFormat command with --tpm to bind a key to the tpm, but the problem is that since there are no tpm libraries loaded into the initramfs, this command fails with driver/TCTI errors

I know this is not an issue with the stack, but if someone knows how to do this, I would appreciate the help. I've tried posting on other forums, but nobody really knows enough about how tpm works to be able to help.

**Comment by diabonas** [comment]
You probably need to add the appropriate kernel module for your TPM and the TCTI library `libtss2-tcti-device.so.0` to your initramfs. How this is done depends on the way you/your distribution generates the initramfs; see the files in [tpm2-totp/dist](https://github.com/tpm2-software/tpm2-totp/tree/master/dist) for how this can be done with initramfs-tools (Debian-based distributions), dracut (Fedora, openSUSE and others) and mkinitcpio (Arch Linux).

**Comment by glassman81** [comment]
Thank you. I'm running Raspbian Buster on a RPi 4. I may try clevis again with clevis-tpm2 and clevis-dracut because I think it does load the appropriate libraries.

**Comment by diabonas** [comment]
Clevis includes the appropriate files in the initramfs, at least the current master branch, for the released version 11 you need to apply https://github.com/latchset/clevis/pull/81 first. However IIUC the Raspberry Pi boot process doesn't use an initramfs by default, so simply installing clevis-dracut won't do anything. You have to build an initramfs by hand and enable it in `config.txt`, see e.g. the [Embedded Linux Wiki RPiconfig article](https://elinux.org/RPiconfig) or this ["Raspbian and btrfs" tutorial](https://feldspaten.org/2019/07/05/raspbian-and-btrfs/). I don't know if an initramfs created by dracut will work, most people seem to use initramfs-tools for this purpose, which is the default initramfs generator for Debian.

**Comment by glassman81** [comment]
Thank you. If initramfs-tools with dracut was installed, could that work?

**Comment by diabonas** [comment]
initramfs-tools and dracut are completely separate tools, so installing clevis-tpm2 will not help if you are using `mkinitramfs` to generate the initramfs. Like I said, for initramfs-tools I would try adapting [`tpm2-totp/dist/initramfs-tools/hooks/tpm2-totp.in`](https://github.com/tpm2-software/tpm2-totp/blob/master/dist/initramfs-tools/hooks/tpm2-totp.in) to your needs.

**Comment by glassman81** [comment]
@diabonas I'm going over this right now. From the following,

#!/bin/sh

PREREQ='plymouth'
prereqs() {
    echo "$PREREQ"
}

case $1 in
    prereqs)
        prereqs
        exit 0
        ;;
esac

. /usr/share/initramfs-tools/hook-functions

copy_exec @LIBEXECDIR@/plymouth-tpm2-totp /bin
copy_exec @TSS2_TCTI_DEVICE_LIBDIR@/libtss2-tcti-device.so.0
copy_modules_dir kernel/drivers/char/tpm

I understand that I would probably take Plymouth out as a prereq, but I'm not sure where variables like @LIBEXECDIR@ and @TSS2_TCTI_DEVICE_LIBDIR@ are defined. I only ask because I may want to copy some other files. Is there a list of these variables somewhere, and where they point?

Also, is it enough that once I have my hook file made, that I copy it to '/usr/share/initramfs-tools/hooks' and then run 'mkinitramfs -o /boot/initramfs.gz'?

**Comment by diabonas** [comment]
> I understand that I would probably take Plymouth out as a prereq, but I'm not sure where variables like @LIBEXECDIR@ and @TSS2_TCTI_DEVICE_LIBDIR@ are defined. I only ask because I may want to copy some other files. Is there a list of these variables somewhere, and where they point?

`@LIBEXECDIR` is just the [standard Autoconf `libexecdir`](https://www.gnu.org/prep/standards/html_node/Directory-Variables.html), but you probably don't need it anyway since you don't want to install `plymouth-tpm2-totp`. `@TSS2_TCTI_DEVICE_LIBDIR@` points to wherever you installed the tpm2-tss libraries, which you can find out using `pkg-config --variable libdir tss2-tcti-device`. These variables are defined in [`configure.ac`](https://github.com/tpm2-software/tpm2-totp/blob/master/configure.ac).

> Also, is it enough that once I have my hook file made, that I copy it to '/usr/share/initramfs-tools/hooks' and then run 'mkinitramfs -o /boot/initramfs.gz'?

For your own scripts you should use `/etc/initramfs-tools/hooks` (and `/etc/initramfs-tools/scripts/init-premount` if you want to run shell scripts in the initramfs) instead of `/usr/share` which is reserved for packages, but yeah, that should be sufficient.

**Comment by glassman81** [comment]
@diabonas I want to thank you for being so patient with me. Your handle has already made it into a powerpoint slide for a presentation in class. In fact, a lot of of the TPM developers on GitHub have been kind enough to respond to various questions I've had, and I'm really thankful for that. I haven't had much luck on other forums.

With that in mind, I wanted to say that I abandoned using clevis to unlock the encrypted root as it doesn't seem to really work. However, I ran into an interesting problem when trying to do it another way.

I decided to load a key directly into the TPM's nv storage, and then unseal it at boot to try and automatically unlock my encrypted root ext4 partition. Here's how I went about it.

As noted above, the root is already encrypted as a luks2 volume using a passphrase. Since I want to unlock this volume automatically at boot with my TPM, I made another key using `dd if=/dev/random of=secret.bin bs=1 count=32` and then bound that to the luks volume using `cryptsetup luksAddKey /dev/mmcblk0p2 secret.bin`.

I then took that key and put it into the TPM with the following commands. I am using TPM Tools 3.2.1 and TSS version 2.2.1:

`tpm2_pcrlist -L sha256:0,2,3,7 -o pcrs.bin`
`tpm2_createpolicy -P -L sha256:0,2,3,7 -F pcrs.bin -f policy.digest`
`tpm2_createprimary -H e -g sha256 -G rsa -C primary.context`
`tpm2_create -g sha256 -G keyedhash -u obj.pub -r obj.priv -c primary.context -L policy.digest -A "noda|adminwithpolicy|fixedparent|fixedtpm" -I secret.bin`
`tpm2_load -c primary.context -u obj.pub -r obj.priv -C load.context`
`tpm2_evictcontrol -A o -c load.context -S 0x81010002`

I next edited the crypttab file, using the following line, appending the keyscript at the end:

`sdcard UUID=443d9351-4e0d-4fc8-b743-b9215b379939 none luks,keyscript=/usr/bin/passphrase-from-tpm,initramfs`

The keyscript referenced from crypttab file above has the following information:

`#!/bin/sh`
`set -e`
`echo "Unlocking via TPM" >&2`
`export TPM2TOOLS_TCTI="device:/dev/tpm0"`
`/usr/bin/tpm2_unseal -H 0x81010002 -L sha256:0,2,3,7 && exit`
`/usr/lib/cryptsetup/askpass "TPM unlock failed. Enter passphrase for $CRYPTTAB_SOURCE ($CRYPTTAB_NAME): "`

Next, I created the hook script inside the `/etc/initramfs-tools/hooks` directory. That file contains the following data:

`#!/bin/sh`
`case $1 in`
`prereqs) exit 0;;`
`esac`
`. /usr/share/initramfs-tools/hook-functions`
`copy_exec /usr/bin/tpm2_unseal`
`copy_exec /usr/lib/arm-linux-gnueabihf/libtss2-tcti-device.so` 
`copy_file script /usr/bin/passphrase-from-tpm`
`copy_modules_dir kernel/drivers/char/tpm`

I then edited the modules file inside the /etc/initramfs-tools/ directory so that the drivers would load from initramfs This is what I added to that file:

`tpm`
`tpm_tis_core`
`tpm_tis_spi`

Finally, I regenerated the initramfs by running `sudo mkinitramfs -o /boot/initramfs.gz`. I thought this would work, but when I rebooted, I could see on the screen that the drivers weren't loaded properly, and I don't know why. These are the messages I receive at the end:

`ERROR:tcti:src/tss2-tcti-device.c:440:Tss2_Tcti_Device_init() Failed to open device file /dev/tpm0: No such file or directory`
`ERROR: tcti init allocation routine failed for library: "device" options: "(null)"`
`ERROR: Could not load tcti, got: "device"`

`TPM unlock failed. Enter passphrase for /dev/mmcblk0p2 (sdcard): `

The confusing thing about this is that when I dropped into an initramfs shell and ran the `lsmod` command, the drivers tpm, tpm_tis_spi and tpm_tis_core were loaded, but when running  the command `ls /dev`, tpm is not listed among the device files. Is there anything I'm missing? I feel like I'm so close now, but this is proving to be a frustrating issue.

**Comment by glassman81** [comment]
@diabonas I would like to thank you for being so patient with me. In the process of capturing TPM related information from the issue tracker, your handle, as well as others, have had the honor of being elevated to the holiest of grails - a slide in a powerpoint presentation. A lot of devs here have been kind enough to respond to various questions I've had, and I'm really thankful for that. I haven't had much luck on traditional forums.

With that in mind, I wanted to say that I abandoned using clevis with initramfs to unlock the encrypted root as it doesn't seem to really work. Instead, I decided to load a key directly into the TPM's nv storage and then unseal it at boot to try and automatically unlock my encrypted root ext4 partition. However, I ran into an interesting problem when trying to do it this way, and the following details the steps I took to get there.

As noted above, the root is already encrypted as a luks2 volume using a passphrase. Since I want to unlock this volume automatically at boot with my TPM, I made another key using `dd if=/dev/random of=secret.bin bs=1 count=32` and then bound that to the luks volume using `cryptsetup luksAddKey /dev/mmcblk0p2 secret.bin`.

I then took that key and put it into the TPM with the following commands. I am using TPM Tools 3.2.1 and TSS version 2.2.1:

`tpm2_pcrlist -L sha256:0,2,3,7 -o pcrs.bin`
`tpm2_createpolicy -P -L sha256:0,2,3,7 -F pcrs.bin -f policy.digest`
`tpm2_createprimary -H e -g sha256 -G rsa -C primary.context`
`tpm2_create -g sha256 -G keyedhash -u obj.pub -r obj.priv -c primary.context -L policy.digest -A "noda|adminwithpolicy|fixedparent|fixedtpm" -I secret.bin`
`tpm2_load -c primary.context -u obj.pub -r obj.priv -C load.context`
`tpm2_evictcontrol -A o -c load.context -S 0x81010002`

I next edited the crypttab file, appending the keyscript and initramfs options at the end:

`sdcard UUID=443d9351-4e0d-4fc8-b743-b9215b379939 none luks,keyscript=/usr/bin/passphrase-from-tpm,initramfs`

The keyscript referenced from crypttab file above contains the following information:

`#!/bin/sh`
`set -e`
`echo "Unlocking via TPM" >&2`
`export TPM2TOOLS_TCTI="device:/dev/tpm0"`
`/usr/bin/tpm2_unseal -H 0x81010002 -L sha256:0,2,3,7 && exit`
`/usr/lib/cryptsetup/askpass "TPM unlock failed. Enter passphrase for $CRYPTTAB_SOURCE ($CRYPTTAB_NAME): "`

Next, I created the hook script inside the `/etc/initramfs-tools/hooks` directory. That file contains the following data:

`#!/bin/sh`
`case $1 in`
`prereqs) exit 0;;`
`esac`
`. /usr/share/initramfs-tools/hook-functions`
`copy_exec /usr/bin/tpm2_unseal`
`copy_exec /usr/lib/arm-linux-gnueabihf/libtss2-tcti-device.so` 
`copy_file script /usr/bin/passphrase-from-tpm`
`copy_modules_dir kernel/drivers/char/tpm`

I then edited the modules file inside the /etc/initramfs-tools/ directory so that the drivers would load from initramfs This is what I added to that file:

`tpm`
`tpm_tis_core`
`tpm_tis_spi`

Finally, I regenerated the initramfs by running `sudo mkinitramfs -o /boot/initramfs.gz`. I thought this would work, but when I rebooted, I could see on the screen that the drivers weren't loaded properly, and I don't know why. These are the messages I receive at the end:

`ERROR:tcti:src/tss2-tcti-device.c:440:Tss2_Tcti_Device_init() Failed to open device file /dev/tpm0: No such file or directory`
`ERROR: tcti init allocation routine failed for library: "device" options: "(null)"`
`ERROR: Could not load tcti, got: "device"`

`TPM unlock failed. Enter passphrase for /dev/mmcblk0p2 (sdcard): `

The TPM unlock failed message at least lets me know that the keyscript is running that I defined in crypttab. However, the confusing thing about this is that  /dev/tpm0 doesn't seem to exist.  To troubleshoot, I dropped into an initramfs shell and ran the `lsmod` command, and just like I expected, the drivers tpm, tpm_tis_spi and tpm_tis_core were loaded, but when running  the command `ls /dev`, anything tpm is indeed not listed among the device files. Is there anything I'm missing? I feel like I'm so close now, but this is proving to be a frustrating issue.

**Comment by diabonas** [comment]
> @diabonas I would like to thank you for being so patient with me. In the process of capturing TPM related information from the issue tracker, your handle, as well as others, have had the honor of being elevated to the holiest of grails - a slide in a powerpoint presentation. A lot of devs here have been kind enough to respond to various questions I've had, and I'm really thankful for that. I haven't had much luck on traditional forums.

:smile: Thanks :)

> With that in mind, I wanted to say that I abandoned using clevis with initramfs to unlock the encrypted root as it doesn't seem to really work. Instead, I decided to load a key directly into the TPM's nv storage and then unseal it at boot to try and automatically unlock my encrypted root ext4 partition. However, I ran into an interesting problem when trying to do it this way, and the following details the steps I took to get there.

Your approach and the scripts you use seem reasonable to me at first glance.

> I then edited the modules file inside the /etc/initramfs-tools/ directory so that the drivers would load from initramfs This is what I added to that file:
> 
> `tpm`
> `tpm_tis_core`
> `tpm_tis_spi`

Usually loading the modules should be handled automatically by udev, so editing `/etc/initramfs-tools/modules` shouldn't be necessary. If you omit these lines and do `lsmod` in the initramfs shell, do these modules get loaded as well? If not, there might be a problem with udev, which is responsible for creating the device files like `/dev/tpm0` (or maybe Raspbian uses a different method here, I don't have any experience with that distribution).

That being said, if you have to load the modules manually, you might be missing some required modules: when the system is fully booted (I assume the TPM works as expected then), what is the output of `lsmod | grep tpm`? On my system, that shows the loaded modules `tpm`, `tpm_tis_core`, `tpm_tis` and `tpm_crb`. The last two aren't included in your list, it might be necessary to add them as well (but that depends entirely on the TPM model, of course).

> The TPM unlock failed message at least lets me know that the keyscript is running that I defined in crypttab. However, the confusing thing about this is that /dev/tpm0 doesn't seem to exist. To troubleshoot, I dropped into an initramfs shell and ran the `lsmod` command, and just like I expected, the drivers tpm, tpm_tis_spi and tpm_tis_core were loaded, but when running the command `ls /dev`, anything tpm is indeed not listed among the device files. Is there anything I'm missing? I feel like I'm so close now, but this is proving to be a frustrating issue.

Is there anything in `/dev` at all, e.g. your storage devices, connected peripherals, ...? If so, at least udev should be working as expected. Also, what is the output of `dmesg | grep -i tpm` when run in the initramfs shell and when run on the full system?

**Comment by glassman81** [comment]
> (or maybe Raspbian uses a different method here, I don't have any experience with that distribution).

To answer your first question, I believe raspbian uses a standard udev implementation.

> That being said, if you have to load the modules manually, you might be missing some required modules: when the system is fully booted (I assume the TPM works as expected then), what is the output of `lsmod | grep tpm`?

Using `lsmod` in the OS only lists the three tpm drivers - `tpm`, `tpm_tis_core`, and `tpm_tis_spi`. After taking the tpm modules out of the initramfs modules file and going back into initramfs like you suggested, none are listed.

> Also, what is the output of `dmesg | grep -i tpm` when run in the initramfs shell and when run on the full system?

A `dmesg | grep -i tpm` run in the OS yields this output:

`[    5.256589] tpm_tis_spi spi0.1: 2.0 TPM (device-id 0x1B, rev-id 22)`

Using the same command in initramfs gives me nothing.

> Is there anything in `/dev` at all, e.g. your storage devices, connected peripherals, ...? If so, at least udev should be working as expected.

There is in fact device files in /dev for initramfs. You got me thinking, so I did a comparison of `/dev` in initramfs vs in the OS. This is the output from initramfs:

`autofs
block
bus
cachefiles
char
console
cpu_dma_latency
disk
dm-0
fb0
fd
full
gpiochip0
gpiochip1
gpiomem
hidraw0
hidraw1
hidraw2
hidraw3
hidraw4
hidraw5
hwrng
input
kmsg
loop0
loop1
loop2
loop3
loop4
loop5
loop6
loop7
loop-control
mapper
mem
memory_bandwidth
mmcblk0
mmcblk0p1
mmcblk0p2
network_latency
network_throughput
null
ptmx
pts
ram0
ram1
ram10
ram11
ram12
ram13
ram14
ram15
ram2
ram3
ram4
ram5
ram6
ram7
ram8
ram9
random
raw
stderr
stdin
stdout
tty
tty0
tty1
tty10
tty11
tty12
tty13
tty14
tty15
tty16
tty17
tty18
tty19
tty2
tty20
tty21
tty22
tty23
tty24
tty25
tty26
tty27
tty28
tty29
tty3
tty30
tty31
tty32
tty33
tty34
tty35
tty36
tty37
tty38
tty39
tty4
tty40
tty41
tty42
tty43
tty44
tty45
tty46
tty47
tty48
tty49
tty5
tty50
tty51
tty52
tty53
tty54
tty55
tty56
tty57
tty58
tty59
tty6
tty60
tty61
tty62
tty63
tty7
tty8
tty9
ttyAMA0
ttyprintk
ttyS0
urandom
usb
vc-mem
vchiq
vcio
vcs
vcs1
vcsa
vcsa1
vcsm
vcsu
vcsu1
watchdog
watchdog0
zero`

Afterwards, I found that the following were missing from initramfs:

`btrfs-control
cuse
fuse
i2c-1
initctl
log
media0
mqueue
net
ppp
rfkill
serial0
serial1
shm
snd
spidev0.0
tpm0
tpmrm0
uhid
uinput
v4l
vcs2
vcs3
vcs4
vcs5
vcs6
vcs7
vcsa2
vcsa3
vcsa4
vcsa5
vcsa6
vcsa7
vcsm-cma
vcsu2
vcsu3
vcsu4
vcsu5
vcsu6
vcsu7
vhci
video10
video11
video12`

Of these, `spidev0.0` caught my eye. I figured that the SPI drivers are missing (which are needed since the TPM is connected to the SPI bus), so I modified the previous hook script to include them. It now looks like this:

`#!/bin/sh`
`case $1 in`
`prereqs) exit 0;;`
`esac`
`. /usr/share/initramfs-tools/hook-functions`
`copy_exec /usr/bin/tpm2_unseal`
`copy_exec /usr/lib/arm-linux-gnueabihf/libtss2-tcti-device.so`
`copy_exec /usr/lib/arm-linux-gnueabihf/libatspi.so.0`
`copy_file script /usr/bin/passphrase-from-tpm`
`copy_modules_dir kernel/drivers/char/tpm`
`copy_modules_dir kernel/drivers/spi`

After that, I rebooted, and success! I didn't have to type in a password because the TPM unlocked the OS.

I wonder if this was the original issue with clevis not working. I stole the hook and boot scripts for clevis that included initramfs support from [here](https://github.com/mtottenh/clevis), but I abandoned that idea when I couldn't get it to work.

My only other question is this. Because I am unsealing the key to unlock the root partition from initramfs, that is pretty insecure, correct? I'm thinking that someone could get into the initramfs shell and just issue the same tpm2_unseal command to get the key. Would clevis being used from initramfs be more secure? Or is it pretty much the same deal?

Again, thank you @diabonas. I have a feeling I'll be adding more powerpoint slides of this experience very soon!

**Comment by glassman81** [comment]
>Glad you got it working! For posterity, do you still need to add the drivers to `/etc/initramfs-tools/modules`, or does it work automatically now?

Interestingly enough, it will not work unless those modules are added to the modules file. I just tried without them and I received the same error I did initially, but obviously for a different reason:

`ERROR:tcti:src/tss2-tcti-device.c:440:Tss2_Tcti_Device_init() Failed to open device file` `/dev/tpm0: No such file or directory`
`ERROR: tcti init allocation routine failed for library: "device" options: "(null)"`
`ERROR: Could not load tcti, got: "device"`

>For the Raspberry Pi I am not sure: unless the Raspberry PI bootloader happens to measure the contents of the initramfs into one of the PCRs which would allow to detect such tampering, I think the only solution is to restrict access to the physical device.

There really isn't an easy way to do this as the boot loader cannot measure the contents of the initramfs into a PCR, but I was reading up on u-boot, which I think does have this functionality. Unfortunately, since the boot loader, kernel, etc, (on the Pi) is on a fat32 partition on the sd card, it may be kind of pointless to try this, other than for some sort of proof of concept, which does sound kind of fun.

>I would recommend disabling access to the rescue shell and extending the PCR values after you have retrieved the secret so that it can't be unsealed a second time.

I'm not how I would disable access to the shell. As it stands, the only way I get in is by temporarily changing a character to an erroneous value in cmdline.txt (on fat32), so when the kernel reads it, it doesn't know where the correct volume is ... then the initramf shell starts. I later change cmdline.txt back in initramfs to change everything back the way it was. The shell is also on the fat32 partition, and I don't know of any way to disable access to it.

As for extending the PCR values, since the hash values change after an extend, wouldn't that stop the OS from automatically loading? With that in mind, I have a unique issue with the Pi the way it is. If I extend a PCR, the hash values zero out after I shutdown, unplug it, and the start it back up. I thought they were supposed to stay, but what do I know? I'm trying to contact Infineon about that to see where that goes.

**Comment by diabonas** [comment]
> There really isn't an easy way to do this as the boot loader cannot measure the contents of the initramfs into a PCR, but I was reading up on u-boot, which I think does have this functionality. Unfortunately, since the boot loader, kernel, etc, (on the Pi) is on a fat32 partition on the sd card, it may be kind of pointless to try this, other than for some sort of proof of concept, which does sound kind of fun.

Yeah, to have a proper chain of trust, this needs to be done by the firmware stored in the [boot ROM](https://www.raspberrypi.org/documentation/hardware/raspberrypi/booteeprom.md), and the firmware needs to be write-protected to avoid tampering.

> I'm not how I would disable access to the shell. As it stands, the only way I get in is by temporarily changing a character to an erroneous value in `cmdline.txt` (on fat32), so when the kernel reads it, it doesn't know where the correct volume is ... then the initramfs shell starts. I later change cmdline.txt back in initramfs to change everything back the way it was. The shell is also on the fat32 partition, and I don't know of any way to disable access to it.

You can use the kernel command line parameter `panic` according to the [initramfs-tools man page](https://manpages.debian.org/stretch/initramfs-tools-core/initramfs-tools.8.en.html) to disable the shell. Of course if an attacker can change the contents of the SD card, they can remove that parameter as well, so you still have to restrict physical access to the Raspberry Pi.

> As for extending the PCR values, since the hash values change after an extend, wouldn't that stop the OS from automatically loading?

No, you extend the value only after you have retrieved the keyfile, so the PCR values still match those you used during sealing. If you need to update the keyfile or seal it to changed PCR values afterwards, it becomes a bit more difficult because you need to know the PCR values prior to extending them (e.g. by temporarily disabling the extension until you have generated the new key). As an alternative, you can store the keyfile in a NVRAM index and use `tpm2_nvreadlock` to disable reading the index after you have opened the LUKS volume, that saves you from having to extend the PCRs.

> With that in mind, I have a unique issue with the Pi the way it is. If I extend a PCR, the hash values zero out after I shutdown, unplug it, and the start it back up. I thought they were supposed to stay, but what do I know? I'm trying to contact Infineon about that to see where that goes.

That is to be expected: the PCRs are not a permanent record that is kept during the lifetime of the TPM, they measure only the current boot process. The idea is that you start with a trustworthy firmware that measures different aspects of the boot process (the contents of the firmware itself, the firmware configuration, partition layout, contents of executed files, ...) into different PCR registers so that you can detect whether anything has changed. If you cannot trust the firmware, no security guarantees about the integrity of the boot process can be made, e.g. it can reset the PCRs completely using `tpm2_pcrreset`.

**Comment by diabonas** [comment]
> On a Raspberry Pi the kernel command line is cmdline.txt. The initial-fs tools says to use the panic parameter in this format, `panic=<secs>`. In my case I'm guessing that secs would be set to 0, correct?

That would be my interpretation as well, yes. The behaviour of this option seems to have been [somewhat buggy](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=751488) apparently, but should be fixed in the latest version of initramfs-tools.

> Oh, so are you saying the extend would happen only after the key is unsealed and read, assuming that after a reboot the PCR values would reset anyway and boot processes are measured again - making the value of those PCR's the same as when the policy was created for the unseal? When exactly would I perform this extend? Would it be a good idea to do it in the same keyscript file right after the unseal is performed?

I would use `tpm2_pcrextend` directly after `tpm2_unseal` in the keyscript, yes.

> Ok, I assume you mean that wherever in the boot process I decided to extend, temporarily stop that.

:+1:

> Again, if I went this route, would I use this tool in some place like the keyscript, right after the unseal is performed?

Same as above, directly after the `tpm2_nvread` in the keyscript.

> What can reset the PCRs completely? Are you saying I could do this and it wouldn't matter anyway since nothing is being measured regarding the boot process, or are you saying something else?

I suspect that the Raspberry Pi doesn't use the TPM at all, so you can directly access the TPM using the platform authorisation. You can test this by using the command `tpm2_changeauth -c platform testpassword` - if that doesn't fail with an "authorization failure" error, you have platform authorisation and can do things like clearing and resetting the TPM. In that case binding against specified PCRs is pretty meaningless in your context because they don't provide measurements of the current system state anyway (you can find out the current PCR configuration using `tpm2_pcrread`, in that case all PCR indices will have the value `0x0000...` or `0xFFFF...`).

**Comment by tstruk** [comment]
Closing this issue.
Feel free to continue the discussion here if move it to tpm2@lists.01.org mailing list if needed.
